contributors/devel/sig-testing/writing-good-e2e-tests.md

```diff
@@ -38,17 +47,233 @@ like the following generates rather useless errors:
 Expect(err).NotTo(HaveOccurred())
 ```
 
-Rather
+Errors returned by client-go can be very detailed. A better way to test for
+errors is with `framework.ExpectNoError` because it logs the full error and
+then includes only the shorter `err.Error()` in the failure message. To make
+that failure message more informative,
 [annotate](https://onsi.github.io/gomega/#annotating-assertions) your assertion with something like this:
 
 ```
-Expect(err).NotTo(HaveOccurred(), "Failed to create %d foobars, only created %d", foobarsReqd, foobarsCreated)
+framework.ExpectNoError(err, "tried creating %d foobars, only created %d", foobarsReqd, foobarsCreated)
 ```
 
 On the other hand, overly verbose logging, particularly of non-error conditions, can make
 it unnecessarily difficult to figure out whether a test failed and if
 so why?  So don't log lots of irrelevant stuff either.
 
+Except for this special case, using Gomega assertions directly is
```

Comment:
I don't think that there is wide consensus on using Gomega, I personally prefer to use stdlib directly, and I know some other contributors too
```
if !strings.Contains(actualStr, expectedSubStr) {
   framework.Failf("expected %s to contain %s",actualStr, expectedSubStr) 
}
```

Reply:
I see you expand on this later

Reply:
And this already shows how hand-written code can be problematic: is the failure message still readable when `actualStr` is empty? When using `%q`, how readable is it for multi-line strings?

The other problem is inconsistency.

I understand that we cannot require the usage of gomega, but I still think that we should encourage it where it makes sense.

Reply:
👍

Reply:
I'd like to write about https://github.com/kubernetes/kubernetes/pull/113298 here, which implies that we need to get that merged first.

Reply:
I've added something about that.